File: stack/stack.py
from node import Node

class Stack:
    def __init__(self, value):
        self.top = Node(value)
        self.height = 1

    def print(self):
        tmp = self.top
        while tmp is not None:
            print(tmp.value)
            tmp = tmp.next

    def push(self, value):
        new_node = Node(value)
        if self.top is not None:
            new_node.next = self.top
        self.top = new_node
        self.height += 1

    def pop(self):
        if self.top is None:
            return None
        top_node = self.top
        self.top = self.top.next
        top_node.next = None
        self.height -= 1
        return top_node

File: stack/cpp/stack.cpp
#include <iostream>
#include "./node.h"
#include "./stack.h"

Stack::Stack() : height(0), top(nullptr) {}

Stack::Stack(int value) : height(1), top(new Node(value)) {}

int Stack::get_height() const { return height; }

int Stack::get_top() const
{
  if(!top) return INT_MIN;
  return top->value;
}

void Stack::print() const
{
  Node *tmp = top;
  while(tmp)
  {
    std::cout << tmp->value << std::endl;
    tmp = tmp->next;
  }
}

void Stack::push(int value)
{
  Node* new_node = new Node{value};
  new_node->next = top;
  top = new_node;
  height++;
}

int Stack::pop()
{
  if(!top) return INT_MIN;
  Node* node = top;
  int value = node->value;
  top = top->next;
  delete node;
  height--;
  return value;
}

Stack::~Stack()
{
  std::cout << "In Stack::~Stack()" << std::endl;
  if(!top) return;
  while(top)
  {
    std::cout << "Deleting: " << top->value << std::endl;
    pop();
  }
}
